# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two(3))

5


In [3]:
print(add_two)

<function add_two at 0x7fddd06617e0>


In [4]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [5]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)
(lambda x: x + 2)(3)

5

#### Saving to a Variable

In [6]:
f = lambda x,y:x+y+2
print(f)

<function <lambda> at 0x7fddccd1c430>


In [7]:
f(3,2)

7

#### Multiple Inputs

In [9]:
# Syntax: lambda param1, param2, etc: expression
l = 3
w = 5
h = 8

(lambda x,y,z: x*y*z)(l,w,h)


120

In [ ]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)


#### Passing a Function into a Function

In [10]:
def apply_func(func, arg):
    return func(arg)

In [12]:
def my_math(x):
    return x*x


In [16]:
apply_func(my_math,5)

25

#### Passing a Lambda into a Function

In [ ]:
# Call the apply_to_each function, pass in a lambda function as f argument


In [18]:
apply_func(lambda x: x*x, 5)

25

#### Returning a Function from a Function

In [28]:
def create_enclosure():
    num = 0
    def inner_func():
        nonlocal num
        num=num+1
        return num
    return inner_func

my_multiplier = create_enclosure()
print(my_multiplier())
print(my_multiplier())
print(my_multiplier())
print(my_multiplier())
print(my_multiplier())

1
2
3
4
5


In [29]:
def create_multiplier(num):
    def inner_func(x):
        return num*x
    return inner_func
my_multiplier = create_multiplier(2)
my_multiplier(4)

8

In [30]:
my_multiplier(8)

16

#### Returning a Lambda from a Function

In [31]:
def create_multiplier(num):
    return lambda x: num*x

my_multiplier = create_multiplier(2)
my_multiplier(4)

8

#### If Statements within Lambdas

In [32]:
def do_something(num):
    if num>10:
        return num * 2
    else:
        return num + 2
    
do_something(5)

7

In [ ]:
# Ternary: value_if_true if condition else value_if_false


In [34]:
def do_something(num):
    return num *2 if num > 10 else num + 2
do_something(5)

7

In [35]:
(lambda num: num *2 if num > 10 else num + 2)(5)

7

In [40]:
# write a lambda function that takes a name and and age and returns
# a string that says  ""[name] is [this many] years old"

(lambda name, age: f"{name} is {age} years old")("Peter", 40)

'Peter is 40 years old'

In [41]:
# write a lambda function that takes a 3 numbers and returns those numbers added together
(lambda n1,n2,n3: n1+n2+n3)(10,20,30)

60

In [49]:
student = {"name":"Random Person", "IQ": randint(1,240)}
print(student["IQ"])

40


In [46]:
from random import randint

my_class = [
    {"name":"Valencia", "IQ": randint(1,240)},
    {"name":"Mason", "IQ": randint(1,240)},
    {"name":"Cynthia", "IQ": randint(1,240)},
    {"name":"Sheridan", "IQ": randint(1,240)}
]

print(my_class)

[{'name': 'Valencia', 'IQ': 120}, {'name': 'Mason', 'IQ': 64}, {'name': 'Cynthia', 'IQ': 157}, {'name': 'Sheridan', 'IQ': 80}]


In [50]:
sorted(my_class, key=lambda student: student["IQ"])

[{'name': 'Mason', 'IQ': 64},
 {'name': 'Sheridan', 'IQ': 80},
 {'name': 'Valencia', 'IQ': 120},
 {'name': 'Cynthia', 'IQ': 157}]

In [51]:
sorted(my_class, key=lambda student: student["IQ"], reverse=True)

[{'name': 'Cynthia', 'IQ': 157},
 {'name': 'Valencia', 'IQ': 120},
 {'name': 'Sheridan', 'IQ': 80},
 {'name': 'Mason', 'IQ': 64}]

In [42]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [58]:
print(my_class)

[{'name': 'Valencia', 'IQ': 120}, {'name': 'Mason', 'IQ': 64}, {'name': 'Cynthia', 'IQ': 157}, {'name': 'Sheridan', 'IQ': 80}]


In [61]:
my_class.sort(key=lambda student: student["IQ"], reverse=True)
print(my_class)

[{'name': 'Cynthia', 'IQ': 157}, {'name': 'Valencia', 'IQ': 120}, {'name': 'Sheridan', 'IQ': 80}, {'name': 'Mason', 'IQ': 64}]


In [57]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [62]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [66]:
def squared(num1, num2):
    if num1 < 10 and num2 < 10:
        return (num1**2, num2**2)
    else:
        return (num1, num2)
    
nums1 = [4,10,3,1,6]
nums2 = [4,1,20,3,15]

new_list =[]
for i in range(len(nums1)):
    new_list.append(squared(nums1[i], nums2[i]))
    
print(new_list)


[(16, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


In [68]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)
squared_nums_map = map(squared, nums1, nums2)
print(list(squared_nums_map))

[(16, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


In [72]:
def squared(num1, num2):
    return (num1**2, num2**2) if num1 < 10 and num2 < 10 else (num1, num2)
         
    
         
squared(4,4)    

(16, 16)

#### Using Lambda's with Map

In [74]:
squared_nums_map = map((lambda num1, num2: (num1**2, num2**2) if num1 < 10 and num2 < 10 else (num1, num2)), nums1, nums2)
print(list(squared_nums_map))

[(16, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


#### In-Class Exercise #1 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [77]:
nums = [4,2,6,8,15]
# [7, 3, 11, 15, 29]
list(map(lambda num: num * 2 - 1, nums))


[7, 3, 11, 15, 29]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [78]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [85]:
bool({"key":123})

True

In [88]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)
list(filter(None, [True, True, False, "", "string", [], [1,2,4,5] ]))

[True, True, 'string', [1, 2, 4, 5]]

True

#### Using Lambda's with Filter()

In [101]:
names = ["Daniel", "Leisa", "Sheridan", "Desiree", "Keanu", "Santa", "Danny Devito"]
def d_names(name):
    if name[0].lower() == 'd':
        return True
    else:
        return False
    
the_d_names = filter(d_names, names)
list(the_d_names)

['Daniel', 'Desiree', 'Danny Devito']

In [102]:
lamb_ds = filter(lambda name: True if name[0].lower() == 'd' else False, names)
list(lamb_ds)


['Daniel', 'Desiree', 'Danny Devito']

#### In-Class Exercise #2 <br>
<p>Using the filter function, create a list out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [107]:
36 <= 24

False

In [103]:
nums = [2,7,4.2,1.6,9,4.4,4.9]

In [108]:
from statistics import mean
below_mean = filter(lambda x: x <= mean(nums) ,nums)
list(below_mean)


[2, 4.2, 1.6, 4.4]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [111]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



In [117]:
list_1 = [2,4,6,8,10]

def subtract_nums(num1, num2):
    return num1 - num2
result = reduce(subtract_nums, list_1)
print(result)

-26


In [112]:
2-4-6-8-10

-26

#### Using Lambda's with Reduce()

In [118]:
the_sum = reduce(lambda aggregator, current_value: aggregator + current_value ,list_1)
the_sum

30

#### In-Class Exercise #3 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [120]:
result = reduce((lambda total, next_num: total*next_num), list_1)
result

3840

In [123]:
test = "x" + "y"
test

'xy'

In [125]:
letters = ["C", "o", "m", "e", " ", "o", "n", " ", "D", "o","w","n"]
full_phrase = reduce(lambda phrase, letter: phrase + letter, letters)
full_phrase


'Come on Down'

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [139]:
def recursive(i):
    if i == 4: #STOP LOOPING -- Base Case
        return
    print("hi!")
    i+=1
    recursive(i)
    

In [140]:
recursive(0)

hi!
hi!
hi!
hi!


#### Implementing a Base Case

#### Writing a Factorial Function

In [143]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

def factorial(num):
    if num <= 1:
        print('factorial(1)=1')
        return 1
    else:
        print(f"factorial({num})={num} * factorial({num-1})")
        return num * factorial(num-1)
factorial(5000)
        

factorial(50000)=50000 * factorial(49999)
factorial(49999)=49999 * factorial(49998)
factorial(49998)=49998 * factorial(49997)
factorial(49997)=49997 * factorial(49996)
factorial(49996)=49996 * factorial(49995)
factorial(49995)=49995 * factorial(49994)
factorial(49994)=49994 * factorial(49993)
factorial(49993)=49993 * factorial(49992)
factorial(49992)=49992 * factorial(49991)
factorial(49991)=49991 * factorial(49990)
factorial(49990)=49990 * factorial(49989)
factorial(49989)=49989 * factorial(49988)
factorial(49988)=49988 * factorial(49987)
factorial(49987)=49987 * factorial(49986)
factorial(49986)=49986 * factorial(49985)
factorial(49985)=49985 * factorial(49984)
factorial(49984)=49984 * factorial(49983)
factorial(49983)=49983 * factorial(49982)
factorial(49982)=49982 * factorial(49981)
factorial(49981)=49981 * factorial(49980)
factorial(49980)=49980 * factorial(49979)
factorial(49979)=49979 * factorial(49978)
factorial(49978)=49978 * factorial(49977)
factorial(49977)=49977 * factorial

factorial(47368)=47368 * factorial(47367)
factorial(47367)=47367 * factorial(47366)
factorial(47366)=47366 * factorial(47365)
factorial(47365)=47365 * factorial(47364)
factorial(47364)=47364 * factorial(47363)
factorial(47363)=47363 * factorial(47362)
factorial(47362)=47362 * factorial(47361)
factorial(47361)=47361 * factorial(47360)
factorial(47360)=47360 * factorial(47359)
factorial(47359)=47359 * factorial(47358)
factorial(47358)=47358 * factorial(47357)
factorial(47357)=47357 * factorial(47356)
factorial(47356)=47356 * factorial(47355)
factorial(47355)=47355 * factorial(47354)
factorial(47354)=47354 * factorial(47353)
factorial(47353)=47353 * factorial(47352)
factorial(47352)=47352 * factorial(47351)
factorial(47351)=47351 * factorial(47350)
factorial(47350)=47350 * factorial(47349)
factorial(47349)=47349 * factorial(47348)
factorial(47348)=47348 * factorial(47347)
factorial(47347)=47347 * factorial(47346)
factorial(47346)=47346 * factorial(47345)
factorial(47345)=47345 * factorial

RecursionError: maximum recursion depth exceeded while calling a Python object

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [3]:
list(range(5))

[0, 1, 2, 3, 4]

In [4]:
def test():
    yield 1
    yield 2
    yield 3
    yield "a"
    yield ["a","b","c"]
    

In [5]:
for thing in test():
    print(thing)

1
2
3
a
['a', 'b', 'c']


In [11]:
test()

<generator object test at 0x7f255971dcb0>

In [7]:
r = test()
print(list(r))

[1, 2, 3, 'a', ['a', 'b', 'c']]


In [8]:
import random

def generate_random_nums():
    while True:
        yield random.randint(1,420)
    

In [9]:
for num in generate_random_nums():
    print(num)

102
315
145
47
15
2
322
50
338
23
389
348
62
342
317
162
255
292
348
390
113
251
276
261
175
303
49
60
14
87
216
21
79
304
355
180
170
94
376
328
357
43
328
54
225
226
105
381
163
260
2
236
62
85
402
323
295
331
411
419
400
176
167
387
415
205
158
359
391
314
79
332
397
245
56
352
375
295
163
203
125
54
301
229
118
355
33
354
113
211
18
255
388
31
308
407
86
76
225
348
339
257
354
161
407
102
296
171
16
230
32
43
94
81
219
124
401
373
340
166
34
178
56
67
21
3
251
57
353
209
389
151
122
212
339
82
254
82
59
61
39
404
36
392
319
252
390
407
138
132
79
246
70
24
128
401
376
353
287
105
348
178
87
354
22
227
291
59
16
347
387
59
311
92
191
287
114
118
381
202
86
42
405
16
89
415
277
38
216
217
315
380
415
205
305
206
178
266
72
279
248
371
196
255
123
115
230
192
287
210
325
351
266
404
345
200
228
420
330
370
279
61
249
5
404
158
349
205
21
396
200
177
386
70
204
25
188
224
210
234
376
226
170
122
376
221
163
46
136
197
369
312
378
30
236
18
117
322
194
311
81
239
129
21
154
187
184
401


372
201
182
420
404
407
302
205
384
125
382
44
339
77
334
71
236
71
270
391
321
57
376
147
317
283
300
280
143
148
380
413
260
412
394
346
316
69
165
385
83
17
399
7
42
168
220
243
39
237
181
63
86
192
419
56
42
30
41
119
393
145
140
418
346
115
298
404
375
239
365
378
247
301
327
40
396
152
27
329
119
23
101
320
232
307
395
135
281
196
129
189
386
183
88
351
359
420
114
147
97
229
133
43
84
97
409
398
8
115
90
113
88
117
323
314
49
36
242
56
320
247
333
29
188
176
383
75
194
323
303
277
121
289
37
102
161
206
167
301
153
215
129
250
23
128
362
366
371
28
9
253
388
363
108
17
63
238
132
248
211
292
331
146
132
72
187
52
124
239
199
60
118
63
141
406
118
335
209
364
340
179
184
35
322
200
376
359
283
101
74
179
133
98
405
414
125
405
211
36
83
3
244
343
295
389
25
362
415
104
296
228
251
79
315
62
75
99
394
328
277
18
295
215
410
345
293
303
152
187
98
353
332
386
246
281
63
68
188
326
335
158
135
208
88
71
406
190
298
155
174
134
316
7
97
416
263
162
244
269
186
401
62
362
24
400
385
3

157
70
310
86
292
401
147
183
26
264
319
121
415
360
139
192
290
287
312
252
370
225
380
185
205
357
413
85
331
36
223
352
82
187
62
66
156
20
408
336
48
106
388
257
366
137
277
38
216
221
411
282
152
25
362
102
240
273
311
139
79
40
290
382
258
347
250
303
344
124
143
230
313
45
339
293
253
143
153
195
86
138
293
44
335
357
392
160
368
251
326
219
362
264
267
118
126
95
397
416
300
313
232
409
382
69
108
120
57
353
169
322
396
370
332
50
290
36
214
189
342
90
161
266
60
325
302
21
93
281
135
330
404
178
355
328
59
201
362
79
344
276
129
46
409
386
248
350
197
185
320
124
375
130
283
125
105
11
294
315
211
164
44
71
194
313
270
309
222
416
419
359
393
235
146
69
150
166
69
84
103
404
267
156
17
242
66
374
404
57
60
332
377
349
197
46
362
321
372
232
161
217
326
213
156
91
184
319
345
106
338
27
372
178
256
192
235
353
294
55
103
331
161
311
313
193
149
314
319
345
11
204
155
164
331
7
304
406
236
218
72
39
123
292
337
34
263
346
139
203
90
23
293
70
6
34
355
82
4
22
222
305
63
44
123
3

44
182
394
309
108
222
48
210
176
418
115
30
250
357
295
396
223
90
159
418
78
406
123
375
36
12
133
197
372
248
128
46
348
359
223
419
116
283
198
280
145
348
402
163
34
398
361
374
133
235
414
63
53
61
295
68
197
163
352
111
393
136
415
30
395
32
64
315
196
12
325
388
404
53
264
210
386
240
324
399
276
360
3
181
71
182
22
132
215
340
4
352
30
334
142
318
258
235
374
325
171
326
98
127
323
154
380
53
66
160
279
325
233
175
325
371
315
227
53
30
49
145
377
202
221
97
35
128
195
191
70
413
299
122
188
231
359
88
386
64
407
271
284
193
237
41
310
285
7
405
27
289
404
269
101
253
91
32
396
279
224
276
137
215
391
120
92
220
314
399
22
221
3
287
4
384
368
286
111
302
287
79
333
72
266
288
188
45
90
90
107
335
141
266
348
24
37
363
12
249
174
208
265
195
65
114
2
231
232
361
365
108
170
88
93
46
413
352
101
293
333
419
263
190
27
204
65
403
148
168
229
112
377
104
67
91
216
183
38
306
99
199
399
315
35
304
68
47
258
376
108
243
230
370
146
405
325
113
1
263
126
305
351
414
359
228
376
99
26

233
407
137
13
248
204
3
268
60
237
320
16
14
42
207
103
417
104
249
120
280
397
300
249
400
8
39
64
371
301
16
289
384
407
252
334
169
342
180
177
191
240
413
131
121
417
412
227
29
318
344
169
229
212
155
246
59
371
192
80
47
161
130
411
382
148
278
114
83
193
417
98
161
343
186
332
246
240
305
125
156
299
266
159
147
220
168
380
144
373
3
2
238
293
165
285
139
363
408
244
118
122
409
352
279
32
362
344
201
410
207
384
49
30
216
53
362
403
200
218
410
303
380
115
208
214
347
13
80
75
378
201
209
1
377
20
357
145
132
38
127
129
86
306
27
34
152
216
19
344
76
141
329
203
329
28
322
230
253
252
140
194
73
29
407
420
194
174
338
20
271
214
80
67
403
235
224
411
281
211
187
293
218
371
58
139
62
129
266
249
240
284
404
78
4
389
12
7
75
340
262
73
284
97
349
360
383
39
266
350
339
378
349
218
197
335
305
297
123
91
312
38
352
410
151
359
251
200
234
8
176
16
355
57
240
409
370
197
361
377
363
386
127
263
177
180
34
223
213
101
126
182
308
383
114
102
94
346
130
37
306
153
309
43
272
378
24

142
35
196
53
246
187
171
281
175
163
125
153
213
220
32
178
355
307
222
94
419
275
197
127
107
412
418
252
6
14
234
341
187
402
75
250
82
259
36
80
97
140
298
318
248
240
228
182
161
181
232
157
24
56
271
319
202
164
410
328
224
118
59
134
168
100
249
83
316
68
80
286
366
82
379
415
263
109
198
129
148
125
418
404
368
214
190
150
69
398
212
152
399
345
68
126
260
218
56
283
380
127
152
202
343
366
259
134
103
188
379
329
7
381
315
225
153
331
305
396
4
244
277
46
417
420
257
172
261
51
68
127
187
255
98
324
212
76
403
354
78
104
245
313
92
34
286
177
402
140
329
406
174
270
116
368
211
179
239
337
17
406
314
56
62
332
258
43
87
170
107
27
73
43
248
337
398
380
301
416
220
134
192
284
211
101
114
130
257
103
195
169
271
332
30
408
241
191
68
129
215
330
69
82
89
34
166
280
337
234
367
416
118
322
126
218
182
264
190
303
270
139
81
158
287
143
342
405
136
148
174
398
169
419
95
352
264
201
164
338
391
394
44
324
129
203
269
114
382
55
309
263
54
89
233
393
313
273
131
101
77
413
227
300

335
351
371
369
286
148
386
225
307
192
105
326
362
206
45
116
393
386
315
294
252
134
308
413
248
244
416
346
217
43
119
284
152
128
3
192
231
231
197
240
312
289
30
18
209
81
226
125
106
25
288
207
253
5
112
308
250
358
137
322
162
164
189
162
366
24
241
151
346
226
335
83
341
362
333
246
333
110
115
48
64
163
140
189
331
392
75
360
82
52
286
344
224
397
105
407
214
143
292
263
186
345
125
407
39
98
162
45
125
353
228
1
331
380
80
200
79
195
56
388
175
250
158
377
54
326
343
52
225
155
288
350
314
138
386
137
27
276
380
312
271
193
118
278
110
227
17
236
135
402
189
333
245
321
205
18
83
238
89
374
72
5
59
244
112
38
121
228
409
56
54
166
371
86
44
95
8
53
404
358
260
118
116
182
54
135
296
45
45
208
285
354
400
326
152
255
119
148
240
23
124
203
407
91
118
308
164
131
328
138
385
10
372
226
218
249
85
226
25
270
17
291
73
184
248
14
303
380
64
285
9
87
370
353
164
216
363
242
235
372
341
50
10
293
327
357
408
129
221
72
201
139
222
384
349
220
130
266
386
365
335
30
264
257
396
20
2

64
88
200
212
68
279
409
3
238
188
412
93
355
236
164
350
392
4
42
73
261
412
190
303
317
182
402
90
409
356
35
200
239
114
412
325
77
34
378
231
395
4
215
181
309
323
55
353
342
302
196
183
282
165
245
199
185
164
42
203
86
420
350
325
121
129
310
397
361
4
17
353
402
40
176
75
162
26
402
75
374
275
104
140
293
344
203
50
185
119
54
380
85
215
66
371
203
204
361
319
92
175
289
156
166
127
82
311
371
230
251
208
33
185
298
237
249
81
228
65
297
412
221
368
113
17
31
297
67
108
161
108
398
128
334
99
46
51
185
109
319
295
278
394
328
128
387
125
27
393
240
390
15
159
158
99
107
402
412
267
61
367
250
132
402
406
217
253
324
406
33
106
328
38
99
245
133
284
331
341
253
1
270
130
199
387
254
248
397
174
361
44
116
52
28
122
199
95
8
199
33
387
411
183
238
103
336
94
285
418
165
195
5
238
158
301
350
74
397
416
30
406
132
227
181
359
3
331
326
88
404
149
364
171
239
306
81
208
199
391
243
329
132
154
164
361
72
262
351
395
268
2
21
212
297
254
324
307
333
371
68
210
28
307
31
159
384
212
3

4
256
311
280
294
108
397
48
410
7
359
198
258
100
249
110
362
253
67
177
93
2
155
38
414
126
236
330
414
4
6
50
153
364
227
169
275
325
11
274
230
51
311
332
277
305
3
149
313
30
337
344
101
44
90
232
208
69
56
113
242
100
95
330
223
9
94
273
186
147
166
311
298
276
398
252
253
199
203
374
332
171
52
23
305
217
189
277
152
380
60
302
275
260
216
324
36
381
22
14
251
344
259
281
370
403
400
21
194
299
308
317
283
369
174
203
152
14
372
110
261
98
77
209
289
316
276
248
131
314
404
6
347
196
160
288
100
410
170
287
256
84
328
211
267
97
35
378
89
41
416
16
173
241
169
318
149
310
388
11
370
323
106
205
222
105
267
38
84
391
75
399
73
288
248
374
184
336
69
28
359
237
183
161
21
142
128
296
338
111
1
280
212
340
388
346
85
394
298
293
330
228
89
397
186
331
336
187
182
400
199
56
111
41
303
77
247
382
35
144
197
22
68
221
85
203
217
88
215
106
385
377
400
162
15
196
204
291
256
261
291
212
384
404
369
299
355
374
329
228
354
31
382
396
219
2
355
233
238
370
7
321
328
12
265
187
29
167
11

161
336
68
363
291
136
404
262
105
248
134
154
47
157
10
162
108
24
183
151
110
78
163
120
148
370
193
406
256
167
83
224
115
131
175
283
34
299
228
114
188
180
140
234
374
84
309
390
416
205
208
280
326
71
347
158
254
157
405
11
359
127
399
75
101
96
8
82
43
229
381
34
51
328
252
306
40
207
335
167
172
388
400
54
16
1
285
224
248
262
314
34
193
241
146
315
296
88
399
273
64
217
374
315
310
106
282
82
184
349
266
240
310
245
132
396
171
76
222
390
360
175
41
360
157
116
42
357
61
229
345
299
112
123
252
306
371
255
79
57
93
278
408
99
229
117
154
71
386
236
76
315
220
193
312
241
81
188
23
234
123
214
103
6
376
92
122
337
52
2
228
205
414
40
337
350
79
126
287
371
348
261
138
289
383
323
211
101
207
21
160
174
225
187
14
146
76
82
156
91
145
7
258
158
401
391
194
222
72
379
339
76
12
132
242
126
101
93
113
394
387
322
286
120
212
275
284
146
338
48
400
30
62
366
255
52
10
324
72
246
399
46
28
325
27
126
102
370
380
309
159
275
48
176
17
249
82
80
380
183
129
43
183
102
257
304
43
86
11

173
287
371
370
76
185
49
387
330
242
275
198
115
105
288
65
278
407
103
159
340
149
98
77
224
111
254
203
264
367
97
60
136
347
278
303
369
212
224
84
325
183
122
387
252
384
335
419
312
87
239
67
390
96
143
198
398
330
88
185
184
45
79
380
157
70
120
348
272
215
124
408
338
320
267
162
304
141
193
27
372
32
394
7
309
21
32
40
406
376
69
78
243
281
79
67
20
195
14
217
370
86
344
160
161
301
390
113
406
93
41
97
143
249
159
388
159
185
77
384
58
231
271
50
366
159
374
70
9
255
249
139
172
295
411
117
37
77
350
351
258
403
211
109
202
205
235
151
416
406
276
170
54
98
112
386
331
17
205
109
303
124
120
293
232
205
300
369
212
394
333
113
280
410
324
3
204
190
41
338
290
410
230
397
242
37
416
134
410
27
260
309
50
300
304
313
320
3
35
413
146
420
21
265
107
288
250
253
327
255
334
79
23
261
407
219
6
52
340
370
29
347
229
257
276
1
168
218
349
356
405
411
268
332
182
57
118
395
412
345
276
146
364
282
397
11
19
155
177
311
401
201
98
33
364
197
142
201
15
92
188
174
177
106
149
57
391
2

31
370
343
303
384
178
293
414
222
104
120
325
231
5
390
73
402
103
168
1
105
211
156
54
135
94
26
371
45
387
54
28
313
403
142
245
127
87
33
23
162
46
20
334
20
394
155
45
24
19
302
308
328
137
309
218
109
259
187
124
283
177
401
234
97
161
285
401
100
187
112
327
45
53
33
285
383
402
33
233
173
9
207
183
109
340
315
61
80
189
213
255
256
63
121
134
156
61
163
167
262
222
214
285
72
281
116
335
74
47
392
121
82
14
133
230
42
208
74
25
378
267
192
260
281
48
150
261
210
171
83
192
420
406
389
205
13
158
87
131
162
9
182
250
246
6
171
178
240
84
267
298
366
207
330
27
208
261
170
204
125
211
353
375
111
14
342
168
50
118
98
323
303
40
239
289
105
166
241
39
63
378
25
49
352
105
354
273
265
312
331
126
420
89
136
411
221
402
249
47
392
398
23
241
1
368
146
38
256
286
282
85
63
264
330
329
220
335
62
196
82
288
414
133
43
259
69
182
253
197
220
294
372
34
233
347
298
6
413
360
145
323
136
197
256
284
213
247
143
155
185
281
97
303
74
275
288
334
119
217
260
406
52
45
135
172
225
179
61
79

332
308
37
339
275
301
11
114
65
380
326
73
61
23
161
282
256
221
52
399
316
409
310
181
79
299
247
104
372
27
100
399
407
380
378
402
190
200
307
108
119
38
339
201
331
168
67
88
369
90
380
420
112
326
190
199
303
167
51
41
185
27
286
91
223
32
413
236
261
158
112
333
286
285
128
95
264
173
352
419
68
32
62
276
45
9
292
245
134
182
82
173
230
361
186
181
138
332
406
393
99
200
52
149
347
371
310
418
103
339
303
402
331
300
369
54
43
60
139
262
24
242
405
175
370
266
234
38
407
385
394
367
417
390
55
1
293
52
277
354
157
191
91
336
137
54
90
131
409
150
23
395
102
29
313
63
264
415
354
135
365
120
6
80
165
19
94
172
212
37
381
90
43
333
26
40
63
14
301
36
147
142
169
379
350
49
174
169
240
249
48
350
248
281
176
214
52
340
144
185
415
293
279
338
78
64
373
121
141
262
188
102
298
198
309
65
241
389
49
98
18
1
398
312
115
126
268
345
147
306
9
312
213
302
59
311
320
205
60
167
304
166
265
93
67
292
413
413
104
39
157
99
226
40
351
29
294
31
169
3
54
285
347
391
281
66
286
349
223
89
159

395
137
357
136
170
97
151
13
206
77
15
265
391
83
32
29
198
172
251
125
32
122
64
300
223
240
51
287
150
417
285
150
238
3
73
127
49
68
116
54
7
55
338
124
104
112
235
96
366
233
186
367
115
11
299
112
339
131
67
292
31
350
223
279
403
224
309
214
304
29
177
146
36
400
391
165
208
32
253
410
121
264
251
66
233
406
281
7
72
381
42
73
358
24
415
396
234
199
193
172
308
25
405
211
322
194
286
262
105
238
249
190
199
68
387
274
69
413
27
298
170
88
169
44
394
384
101
15
111
156
347
195
143
22
223
5
342
310
206
194
251
226
8
311
93
311
150
179
239
31
258
213
74
77
218
140
246
120
195
263
174
99
32
297
2
308
272
419
251
209
125
407
112
391
240
213
138
166
234
72
122
80
228
330
165
144
84
271
31
45
245
212
195
363
148
32
34
140
114
405
301
303
394
255
282
61
95
4
117
67
11
394
50
342
267
144
212
306
102
202
93
344
244
208
255
111
362
350
234
116
11
266
305
359
377
347
47
4
323
136
80
171
186
15
65
346
82
67
19
393
21
295
411
97
20
33
315
123
94
244
383
370
165
319
261
192
290
215
405
109
182

61
50
294
368
376
417
355
354
90
298
190
342
257
87
50
72
199
86
104
298
174
366
223
311
198
125
233
59
142
15
351
96
257
59
230
51
62
140
223
124
290
220
284
123
199
310
380
77
338
95
55
118
157
170
180
204
218
314
278
215
189
223
170
187
151
252
339
255
32
408
98
219
12
314
162
313
45
56
352
396
401
9
300
155
176
283
29
245
184
253
185
306
64
404
354
393
348
228
86
406
110
338
189
190
268
314
276
311
22
390
18
95
386
122
250
28
286
367
49
368
334
290
290
285
247
380
242
420
169
29
54
167
43
312
264
291
51
190
10
123
195
258
145
252
233
416
301
327
297
280
404
15
174
166
209
35
331
301
100
393
364
364
131
225
306
84
301
166
221
41
363
118
317
233
38
142
13
351
63
80
287
311
186
370
162
32
393
289
163
371
344
305
395
26
40
26
127
178
292
212
37
59
97
160
326
258
346
232
267
152
229
290
159
371
292
190
32
419
71
170
143
367
371
420
412
344
74
91
138
162
200
420
326
388
266
171
161
25
149
384
258
185
205
168
365
290
245
258
369
117
103
150
34
213
27
46
50
387
412
174
307
176
370
59
94
36

163
39
129
181
260
47
77
326
113
301
388
96
351
301
296
362
120
176
67
365
159
172
262
26
253
22
140
400
254
410
223
237
235
420
79
406
374
93
207
369
298
370
63
173
307
367
246
394
388
323
124
262
48
236
359
234
181
33
205
300
302
73
256
97
30
321
378
138
211
349
317
103
280
362
44
288
388
171
205
134
248
170
325
327
328
165
258
317
335
321
129
79
406
10
300
259
78
192
142
148
178
269
3
196
389
104
114
136
256
291
211
305
348
243
196
387
65
249
50
282
278
262
168
114
347
107
127
375
84
82
182
34
188
143
420
158
391
338
228
30
179
295
260
44
315
264
137
305
8
278
125
177
144
286
60
241
296
99
32
4
169
98
128
193
28
46
42
183
8
120
145
47
246
99
269
5
388
129
162
9
241
83
131
154
397
102
368
70
389
304
311
263
192
322
395
3
41
307
56
114
351
73
378
262
369
135
417
200
294
24
144
219
133
137
52
136
207
405
236
102
253
324
169
205
278
216
240
401
345
110
317
172
41
42
231
58
259
7
201
358
328
191
240
173
298
109
100
148
277
242
198
250
320
107
15
43
111
289
364
296
229
357
258
127
164
372

361
96
141
173
22
316
223
242
191
397
256
197
201
5
140
265
405
239
64
348
63
331
280
198
99
416
15
75
266
157
151
103
203
161
194
243
178
314
228
240
245
212
266
66
55
295
420
369
328
162
315
96
279
301
18
247
301
81
116
170
33
120
275
96
381
188
44
236
267
75
339
142
349
282
382
70
152
171
93
418
147
299
302
185
3
184
210
263
250
108
134
65
214
65
320
308
416
10
134
73
336
389
117
341
155
122
54
347
413
229
369
323
412
285
117
279
117
140
23
234
146
78
66
259
314
151
314
84
397
382
113
48
113
1
151
110
113
351
347
313
91
140
175
48
138
175
412
177
206
14
120
81
35
143
159
209
139
21
188
413
47
5
37
59
419
175
340
238
140
198
101
157
311
406
361
416
79
176
110
360
241
254
358
338
61
112
293
164
336
131
112
420
222
347
319
3
111
418
171
379
215
274
63
152
184
21
82
64
362
217
82
285
87
210
31
399
224
170
252
170
15
31
295
232
196
32
8
183
370
130
369
272
386
128
128
246
119
323
261
185
10
356
252
172
348
139
39
413
177
295
61
387
228
127
369
94
330
21
343
266
151
166
274
162
135
34
399

34
316
418
111
206
19
20
379
253
239
231
112
399
392
162
383
123
299
262
315
4
318
108
152
410
110
299
410
265
6
144
21
173
261
54
140
271
375
133
332
115
36
262
348
31
9
265
148
138
2
197
369
125
134
189
380
99
202
254
28
301
409
190
35
396
55
255
79
149
186
307
22
39
22
140
187
208
80
273
252
366
31
215
22
84
310
305
79
196
208
320
79
162
336
55
52
374
104
121
134
341
373
211
152
84
23
161
78
39
177
304
319
169
418
25
127
70
72
381
218
122
330
169
61
89
58
260
390
135
102
298
257
308
43
9
261
204
293
374
380
24
259
72
357
222
109
5
278
207
228
245
356
91
329
398
112
59
233
28
402
46
263
405
398
328
147
236
389
402
319
327
206
92
115
53
383
165
348
324
98
319
101
79
102
321
33
332
48
354
134
246
244
372
28
401
73
320
286
81
313
409
156
155
353
99
325
114
272
126
17
134
194
171
129
407
315
261
235
334
32
199
74
370
82
62
288
155
59
216
97
310
205
172
417
258
320
299
6
264
227
140
387
148
414
167
275
193
272
323
166
154
56
60
322
420
193
253
329
324
399
79
389
77
62
416
334
410
237
15
4

81
163
264
97
52
89
353
389
59
159
353
281
11
24
236
43
169
411
154
405
391
325
215
244
396
320
147
300
304
122
354
358
381
231
82
15
61
170
150
37
230
125
208
402
380
204
29
90
247
267
13
39
390
350
173
414
315
32
28
57
174
141
169
200
102
319
357
111
82
216
165
267
294
108
370
21
76
369
183
24
194
50
385
281
234
69
159
267
119
126
237
152
208
51
395
339
254
28
302
133
416
62
324
172
191
399
328
226
57
351
355
397
183
288
155
120
205
174
31
87
412
26
322
416
190
168
298
353
204
222
14
383
264
418
364
413
74
352
238
154
399
33
343
261
180
315
183
53
108
54
393
102
89
171
192
334
120
241
254
126
143
397
328
269
83
129
44
243
364
24
62
116
207
262
88
163
388
200
397
50
217
361
232
26
143
70
125
105
232
352
316
317
285
12
144
384
326
38
112
405
179
205
147
234
337
306
181
155
127
118
376
69
402
226
196
62
272
379
44
46
108
244
120
20
269
173
162
40
256
71
397
24
408
136
253
383
191
33
357
88
345
45
206
83
4
266
273
331
167
162
72
101
210
202
201
149
394
19
404
258
2
86
179
12
104
169
249


98
297
187
174
69
253
173
59
411
215
67
368
356
198
193
350
211
76
320
417
182
45
7
175
86
138
301
301
408
122
157
313
135
14
107
207
8
291
310
274
296
51
55
360
227
70
281
225
21
27
282
329
353
162
367
152
351
30
358
217
308
172
48
40
9
280
364
318
258
23
52
162
148
41
376
14
329
23
208
411
330
248
337
100
213
325
167
188
417
12
117
294
176
172
259
289
130
63
350
202
212
57
390
230
274
209
373
191
99
414
55
412
100
230
288
255
77
351
112
282
185
405
292
91
301
310
392
372
379
310
55
215
187
84
275
196
237
288
54
27
12
81
419
204
411
284
82
28
258
6
245
198
254
197
333
106
154
137
227
91
162
322
8
48
218
172
362
38
148
298
297
396
185
277
190
193
196
367
286
249
126
380
304
358
173
185
2
107
308
173
182
380
214
135
13
160
304
374
133
257
146
137
220
64
377
411
191
187
99
362
169
281
187
92
285
372
127
150
90
358
50
200
136
348
141
92
303
315
353
209
245
155
301
89
177
103
86
33
2
8
18
191
320
299
6
368
272
289
385
38
103
157
320
165
327
304
327
224
28
169
235
313
154
257
24
19
183
417


90
65
260
332
197
229
37
235
38
356
208
324
5
112
140
125
229
314
245
245
189
299
40
15
87
274
376
74
280
243
8
21
227
305
406
351
379
179
354
236
189
398
206
228
194
266
294
99
167
367
67
358
127
412
175
245
300
3
49
134
413
206
331
286
316
27
19
371
155
150
132
59
93
418
347
243
326
141
8
298
387
373
281
249
269
184
121
394
123
345
188
210
9
279
139
280
380
240
166
255
344
15
350
172
4
13
133
254
104
287
126
134
356
170
91
167
25
92
369
412
291
176
210
95
364
344
279
262
404
368
119
273
323
411
14
86
188
316
97
307
385
381
56
297
85
194
225
289
201
180
244
66
127
357
8
391
70
410
396
20
22
129
369
402
19
290
295
171
272
233
17
346
3
238
102
230
265
25
210
357
97
420
336
407
115
145
9
56
388
362
247
236
36
307
176
360
33
193
55
87
110
154
275
32
370
89
394
363
366
127
263
308
392
29
330
368
349
115
301
361
203
18
347
62
202
348
402
229
369
263
204
138
310
1
9
55
108
17
319
207
296
270
147
173
152
73
405
304
295
19
193
142
10
183
217
117
196
242
315
192
358
273
303
270
384
137
104
78
3

180
93
349
42
368
334
291
197
183
109
69
174
54
296
97
405
123
407
169
58
81
155
282
50
3
363
72
42
38
11
250
366
236
107
193
17
30
376
340
82
140
371
206
328
293
42
373
280
347
269
361
11
7
265
392
233
62
260
248
95
3
244
410
131
265
337
374
377
325
35
137
148
322
140
186
274
173
43
374
77
39
354
372
151
313
246
178
80
89
195
115
249
211
285
152
121
140
95
287
58
402
229
232
259
211
352
220
325
72
140
48
319
9
136
127
247
240
416
236
277
131
117
97
49
367
321
123
287
31
305
399
229
211
219
370
37
297
90
348
147
259
121
62
381
280
253
393
226
385
236
159
216
183
168
113
405
13
119
107
36
290
8
152
225
268
151
177
127
327
377
45
327
233
88
171
202
403
255
372
189
418
117
18
4
131
351
396
375
130
89
330
46
133
173
416
119
287
291
158
253
118
341
334
210
407
126
130
343
53
257
30
226
367
258
307
163
325
122
198
410
42
74
63
124
96
286
282
109
1
191
374
79
27
71
415
318
93
97
260
142
40
77
118
253
393
4
187
387
232
160
28
376
416
189
2
128
309
82
65
116
257
369
37
9
357
73
225
197
33
42
33

73
148
412
75
65
230
249
165
281
355
288
56
284
133
117
404
193
16
75
152
13
197
240
370
53
326
175
85
395
75
152
283
416
245
257
173
83
265
192
38
237
104
101
238
46
156
278
397
6
251
58
412
64
119
365
333
128
315
81
407
385
143
289
18
209
211
188
122
164
391
264
137
185
339
207
242
228
368
270
405
188
347
51
148
155
133
300
108
186
53
289
174
309
176
62
416
153
248
365
247
313
129
127
48
116
11
294
316
119
40
94
194
15
352
166
387
347
267
229
217
370
2
72
83
400
268
290
399
404
217
103
65
242
307
327
377
239
176
297
298
122
140
125
180
114
203
344
128
127
263
112
271
190
414
229
277
412
131
132
184
216
310
248
159
144
109
93
77
349
151
290
380
251
112
375
55
246
325
263
119
91
159
62
130
378
184
338
18
21
228
11
347
88
335
143
339
105
211
192
98
303
85
1
15
79
107
270
252
204
265
319
245
117
286
415
4
145
177
374
400
373
376
263
60
254
168
385
333
212
25
278
2
202
259
94
14
58
323
386
116
318
239
155
208
25
236
30
57
314
39
308
211
108
289
48
312
419
67
196
21
70
67
254
182
188
110
2

404
49
198
223
324
50
217
111
327
373
99
244
57
241
354
305
282
86
316
63
253
87
299
208
26
392
406
48
283
410
66
199
177
60
78
215
33
255
56
256
108
141
218
37
13
226
363
374
116
203
348
81
399
239
68
349
272
351
319
295
94
180
313
406
18
388
5
395
67
99
57
146
91
399
324
404
66
96
114
117
132
330
211
38
201
328
78
3
1
245
53
367
7
78
14
81
146
103
169
96
342
285
298
361
48
290
46
187
273
295
59
185
369
64
83
328
367
343
116
163
54
297
75
40
246
299
242
96
79
277
313
109
314
212
85
291
283
102
350
178
151
354
361
347
37
292
147
148
102
96
275
212
211
407
143
218
67
379
388
125
381
390
151
75
269
203
194
149
37
134
262
408
35
187
169
65
55
305
206
71
285
250
84
405
410
116
329
73
279
58
83
212
185
155
372
193
376
38
390
367
112
259
156
257
165
239
327
13
195
63
224
90
228
415
194
385
174
7
230
127
228
286
41
310
274
355
391
167
258
371
1
225
63
157
101
262
34
259
134
246
194
224
31
82
45
169
353
205
144
31
277
373
245
51
7
408
104
99
187
416
24
274
129
76
53
133
257
87
3
299
324
15
42


386
314
315
87
343
190
102
143
369
227
83
86
24
334
305
376
415
231
19
402
409
238
192
21
207
327
277
327
17
92
262
188
371
74
408
229
42
320
317
110
76
157
287
351
353
385
50
255
327
89
84
240
406
384
284
38
89
400
202
146
383
149
409
347
230
239
293
301
21
369
215
163
6
294
176
217
300
27
166
403
260
250
119
128
289
336
324
22
36
302
226
306
114
328
34
414
205
334
370
278
227
301
149
33
38
393
227
344
138
108
141
29
191
336
249
92
349
310
349
254
319
123
183
230
179
154
85
7
360
384
109
349
118
285
71
178
333
374
223
347
1
5
170
188
66
333
220
301
219
282
36
2
288
271
27
287
365
242
49
307
12
262
361
219
354
70
37
138
220
16
208
154
377
181
342
340
27
303
250
166
30
251
246
414
323
73
149
50
233
271
76
336
311
84
240
386
285
321
263
255
29
75
118
13
266
79
370
23
337
366
236
262
334
295
22
237
161
119
154
299
298
226
93
96
385
324
44
367
179
193
161
222
292
61
44
389
181
260
98
224
20
169
398
334
29
269
51
118
4
59
406
86
368
312
21
169
134
182
173
342
346
4
184
132
33
337
246
221
84

26
231
134
191
322
186
297
218
69
103
331
413
132
201
327
258
245
372
2
260
340
21
262
82
272
113
319
182
141
234
203
174
70
99
158
381
28
133
118
349
172
367
232
10
189
64
115
215
160
229
417
377
335
127
226
380
282
363
79
193
204
115
306
211
364
321
27
60
107
54
239
22
140
378
263
19
126
272
290
327
338
371
320
15
219
10
270
171
268
173
116
407
149
248
46
359
380
285
182
329
193
89
7
411
415
124
17
390
324
252
346
273
20
223
5
82
301
394
12
189
375
225
383
38
419
338
63
142
16
203
355
140
395
358
181
225
193
391
102
285
25
136
278
144
342
420
152
332
258
363
7
201
206
318
191
312
21
28
176
85
82
67
1
220
68
368
356
207
120
383
277
30
28
65
119
177
363
386
26
83
270
171
194
231
322
39
271
17
416
409
140
365
257
181
136
116
414
2
357
217
277
143
48
158
36
98
149
376
313
32
293
291
386
77
226
145
30
265
189
359
349
400
329
107
357
54
349
372
38
67
285
353
307
290
235
376
22
67
277
88
14
184
47
152
252
157
85
119
21
144
69
182
312
74
116
340
257
110
388
125
231
419
124
177
236
242
389
29

225
294
380
85
4
81
188
145
328
190
54
414
204
199
8
352
96
406
23
3
283
19
272
7
277
8
382
412
116
81
297
384
357
307
48
243
232
345
57
232
365
133
310
265
164
144
13
307
302
120
60
321
15
73
342
68
226
269
266
194
222
172
293
212
96
5
188
196
310
286
226
349
297
396
221
361
390
256
263
66
169
247
298
362
201
274
254
112
339
80
89
223
27
260
94
168
308
334
62
372
192
322
135
293
66
65
58
145
245
189
12
343
302
74
284
15
255
356
40
297
286
184
377
308
70
17
191
321
8
329
136
362
302
160
189
3
171
312
140
107
371
345
381
368
277
41
198
207
85
336
276
203
66
57
103
11
359
389
104
171
236
19
283
54
273
406
209
356
23
346
219
297
11
163
216
73
34
295
64
281
55
90
153
171
361
182
205
70
271
45
408
393
69
310
386
183
313
351
191
205
161
328
29
355
222
153
16
92
326
121
168
266
275
296
207
396
253
271
160
35
108
359
357
211
15
197
360
168
319
19
148
197
381
368
82
375
192
341
159
53
320
87
361
303
402
402
244
296
346
187
359
203
100
377
73
167
112
137
167
234
329
363
361
375
200
82
139
301
39

24
278
151
365
352
365
223
17
5
132
206
296
133
86
195
142
381
25
63
49
130
100
189
23
211
263
378
295
388
95
249
118
248
302
40
149
316
5
344
230
68
175
50
308
56
105
276
18
299
307
31
249
419
84
138
340
168
117
393
259
336
404
80
323
256
350
319
192
411
42
54
134
116
4
4
192
398
286
198
33
86
366
325
218
96
123
391
52
373
395
181
40
417
251
185
204
48
300
275
325
244
295
176
99
216
393
290
174
290
119
292
293
2
244
134
333
411
265
335
82
309
405
293
372
64
31
265
197
132
167
14
140
258
226
90
92
2
44
11
22
300
347
311
16
323
289
89
67
211
352
369
214
337
348
348
281
219
182
257
106
102
129
318
242
45
201
329
122
160
210
63
364
372
176
103
386
99
210
331
318
71
73
249
298
238
118
287
72
191
217
50
39
257
100
69
393
281
253
283
392
30
5
353
69
232
296
218
264
176
58
33
384
218
111
357
301
268
282
367
204
8
49
217
151
251
305
408
41
165
84
128
228
79
153
260
25
47
241
397
166
145
54
325
249
7
70
140
376
407
209
419
295
414
28
374
228
73
38
163
128
146
287
65
217
363
294
290
246
49
290
3

79
413
357
350
157
398
242
127
156
271
388
149
184
76
77
327
143
246
293
396
227
118
352
169
305
209
258
411
66
281
388
123
363
162
58
195
81
14
93
27
26
200
266
218
234
144
206
322
301
333
21
37
352
51
236
127
141
21
348
74
383
280
414
371
220
57
12
141
55
65
17
162
176
384
48
190
152
76
268
310
392
39
336
381
286
70
345
190
274
145
266
21
4
158
283
366
202
182
224
220
104
417
120
85
420
319
390
9
241
86
274
408
341
238
68
88
141
90
35
16
118
368
91
281
128
275
378
309
260
351
84
287
25
351
104
315
182
293
59
179
84
218
333
229
219
370
227
319
314
341
30
362
352
371
240
305
384
343
124
392
213
31
110
366
210
363
158
263
88
222
342
311
417
336
325
345
323
32
61
34
372
232
409
82
395
64
67
219
341
272
196
413
18
60
179
316
245
45
14
54
124
43
122
12
159
199
345
314
140
331
101
178
186
23
355
27
72
351
310
251
273
244
320
364
232
202
137
206
202
324
194
84
116
291
341
204
44
276
63
211
413
388
354
376
348
363
183
203
356
365
135
14
310
336
196
302
371
110
184
377
266
18
343
402
308
143
3

128
125
288
262
278
12
165
286
65
273
342
138
244
384
93
386
199
412
281
200
40
388
155
36
325
27
86
84
396
267
49
83
227
172
360
202
199
403
181
276
72
252
325
153
28
320
87
248
178
192
326
372
78
51
341
206
130
413
214
109
221
102
208
323
15
13
316
188
71
408
38
315
300
418
31
398
158
77
206
81
250
312
410
19
106
303
184
200
3
215
285
151
385
270
249
217
176
181
264
394
31
402
2
102
362
367
9
162
352
109
201
356
222
4
282
319
231
44
15
82
391
150
179
353
278
224
362
216
101
387
327
283
335
38
248
254
7
294
60
161
282
55
61
20
57
219
243
259
150
86
41
111
417
198
279
238
367
32
186
360
384
416
153
67
357
217
393
367
236
418
14
217
334
347
262
201
294
14
247
248
364
153
27
313
106
119
150
95
257
141
401
354
367
362
88
169
204
44
276
69
393
200
66
324
78
157
181
311
83
27
146
328
165
62
118
275
272
171
131
371
209
365
239
378
68
407
310
5
187
354
338
369
236
56
273
189
355
285
314
263
128
173
389
78
283
53
175
391
29
213
2
275
402
141
238
213
107
16
15
358
321
310
114
334
139
222
229
29

294
98
211
113
209
150
308
156
161
415
184
194
223
298
144
154
174
317
302
111
111
165
145
17
146
82
142
182
407
277
154
237
141
52
191
96
8
72
294
125
71
324
129
26
129
202
410
77
100
417
392
143
59
341
333
379
411
391
239
158
361
253
370
23
385
151
371
303
411
293
90
139
126
226
162
207
306
87
398
247
157
297
300
69
117
142
183
396
213
297
86
82
60
131
26
195
413
391
108
172
184
235
195
246
343
313
213
108
43
404
322
103
417
104
165
246
97
5
255
402
87
127
4
219
96
229
90
357
234
358
84
357
382
175
78
281
217
50
133
251
148
226
276
247
252
416
334
133
300
400
355
7
311
202
110
214
166
17
321
29
371
251
100
365
116
146
234
354
112
257
210
305
73
13
206
223
17
413
370
245
269
184
282
124
17
269
41
56
28
345
420
393
39
269
323
171
419
310
127
160
51
74
251
365
49
279
6
192
199
413
93
51
227
391
393
363
74
370
92
14
44
380
6
50
240
331
3
60
396
331
46
26
361
88
289
256
2
62
168
407
124
127
8
231
376
291
250
118
126
410
388
4
98
300
188
246
161
205
134
11
419
372
276
190
230
313
225
397
2

55
38
260
91
375
351
250
116
131
329
200
400
73
66
107
256
392
402
69
327
85
47
327
61
103
355
137
321
112
365
104
384
180
56
74
141
39
253
397
388
250
224
378
360
130
151
130
57
191
151
254
24
328
398
322
267
217
392
52
120
209
26
232
78
55
417
195
383
175
246
413
250
207
86
284
105
329
327
89
129
341
314
371
256
402
237
108
417
11
361
74
315
66
65
205
162
195
207
97
152
224
413
237
62
235
111
146
158
23
117
122
364
315
83
190
256
358
274
380
70
164
418
63
224
364
124
180
362
301
22
419
130
367
208
5
18
257
316
187
333
91
411
24
285
211
376
173
295
170
325
77
314
74
201
236
333
237
352
380
198
241
32
238
83
204
119
352
58
213
234
137
23
420
70
362
90
35
358
315
258
256
41
41
22
296
16
383
175
59
414
389
228
184
396
142
223
344
47
343
46
401
284
374
127
44
407
236
356
107
227
133
246
116
202
150
29
299
203
94
134
265
243
162
86
4
344
407
32
22
98
161
82
75
174
95
340
151
344
58
385
70
113
130
25
108
358
328
302
253
118
47
65
214
16
52
59
34
124
311
354
55
393
38
177
111
183
324
116
351

201
105
151
160
58
360
120
168
22
29
254
408
41
115
198
126
228
55
382
228
103
387
279
129
234
387
67
225
62
414
157
384
53
264
131
317
142
232
116
352
384
214
113
260
150
183
65
286
329
237
169
389
58
108
412
250
420
362
239
267
200
298
329
56
330
116
401
257
303
186
310
162
187
50
208
134
381
163
172
48
311
228
43
51
293
359
203
49
348
266
144
38
361
334
12
121
326
2
246
152
137
376
225
95
102
177
255
355
130
169
365
269
50
381
95
101
14
9
76
236
52
145
45
33
129
101
170
175
10
111
21
420
345
328
403
22
83
218
272
418
233
255
58
14
314
149
185
5
278
129
167
364
225
204
72
199
162
154
243
330
260
384
373
340
221
220
150
22
285
381
182
387
251
350
134
225
118
411
15
393
28
172
171
19
248
139
74
176
220
156
404
233
257
81
157
102
180
368
122
232
349
22
287
185
342
59
37
375
198
365
304
113
201
266
194
388
297
67
247
26
146
378
201
68
27
165
38
74
27
47
351
129
219
269
197
173
253
58
175
303
162
390
348
310
333
319
404
292
342
297
298
59
213
290
330
77
157
365
368
380
30
275
53
275
232
2

271
348
219
317
140
377
219
63
352
40
112
410
74
162
25
55
340
178
300
415
370
338
220
153
52
216
197
289
418
357
326
181
11
170
35
366
336
282
358
36
57
158
359
398
171
388
154
247
125
106
169
127
119
147
25
253
315
362
18
175
190
51
181
401
179
251
354
292
267
30
100
363
61
140
325
157
32
358
84
44
337
405
394
190
344
46
52
226
289
155
372
247
149
77
366
118
260
294
23
396
159
412
46
355
239
232
245
149
162
87
257
280
342
175
146
50
275
148
212
139
34
257
132
187
88
175
374
132
245
276
30
231
130
238
95
199
295
125
4
248
231
364
26
269
172
407
53
64
50
4
195
196
356
416
291
78
377
381
372
323
250
279
233
70
215
326
178
155
126
57
385
289
377
312
233
393
410
220
320
25
52
355
52
248
69
63
109
200
155
262
413
66
112
224
389
367
406
47
169
164
224
191
128
327
115
139
105
402
8
176
290
366
142
218
193
336
336
413
408
33
405
281
283
206
403
245
411
263
172
260
317
203
28
361
418
150
115
177
420
95
322
292
186
390
30
11
316
135
321
402
259
130
259
245
101
77
133
367
393
204
23
45
150
172
3

369
357
417
309
122
413
388
397
62
138
396
55
27
397
382
354
222
390
279
374
124
89
317
293
144
98
64
105
238
372
125
401
156
361
299
382
206
107
400
108
98
77
272
285
326
397
190
296
386
280
32
203
327
348
18
325
175
99
42
220
389
267
292
379
19
224
228
410
268
89
378
226
78
187
300
72
271
415
15
249
106
171
32
183
187
336
325
131
264
406
174
260
356
413
239
255
338
323
339
108
384
345
278
411
322
414
250
66
124
181
385
114
120
230
58
352
404
339
407
70
83
165
239
213
66
171
77
40
145
152
276
290
53
54
27
342
319
106
178
294
75
156
94
133
154
17
375
88
311
231
55
177
346
216
410
287
253
54
91
347
402
359
254
96
378
237
360
401
112
265
110
232
398
327
399
201
142
87
272
296
75
323
114
163
105
203
221
403
363
357
232
268
208
176
136
417
180
65
325
79
57
143
386
415
414
100
17
99
34
258
105
372
215
52
79
86
204
258
53
29
275
175
230
411
355
79
36
292
385
273
244
145
337
179
208
397
416
402
248
251
194
383
268
331
354
87
380
380
294
130
320
103
248
215
234
258
268
147
155
83
161
344
260
2

289
56
33
299
260
354
50
361
176
283
96
112
143
12
317
114
274
276
247
243
384
268
159
205
54
88
412
49
59
365
6
413
144
384
201
123
362
360
143
338
50
230
344
219
77
245
100
53
53
419
210
301
100
40
165
200
47
151
369
206
360
17
390
134
205
277
390
28
37
143
96
186
68
88
299
78
319
37
13
322
126
173
290
123
222
190
293
177
103
210
355
20
271
136
73
347
75
238
108
369
241
265
259
76
311
303
84
149
37
39
188
187
243
110
312
63
12
238
353
109
196
416
228
16
190
63
351
398
253
154
196
102
381
151
307
74
80
343
144
223
46
95
347
420
386
274
375
28
217
419
207
344
313
37
38
119
65
38
347
307
127
152
400
289
101
43
132
196
189
278
333
1
216
351
403
262
408
291
382
79
178
281
66
281
277
276
344
64
227
249
123
409
410
23
244
381
187
341
123
131
137
206
286
353
184
129
334
21
296
370
285
196
79
236
52
12
203
330
372
382
346
344
349
37
50
79
60
305
361
248
1
83
413
64
244
74
290
412
417
226
330
92
100
68
144
398
242
58
52
19
96
403
389
400
229
79
193
4
305
186
75
248
288
379
23
150
180
255
59
20

342
103
133
296
93
249
53
258
210
354
330
264
361
419
190
13
104
255
350
359
169
208
325
79
70
406
174
273
252
239
79
266
21
298
182
338
342
88
2
286
125
399
33
387
242
116
210
234
293
399
194
79
194
153
238
163
273
419
235
245
21
96
414
4
168
80
171
213
163
169
408
308
225
267
204
243
309
201
93
272
216
400
229
323
293
140
245
160
288
189
400
414
109
42
368
258
89
301
194
141
383
245
309
110
282
12
231
351
333
387
407
159
286
42
1
232
235
242
388
239
23
355
123
249
331
205
328
74
415
169
277
95
258
38
391
388
176
391
179
143
394
315
323
296
412
371
18
386
256
296
25
101
235
87
196
121
343
30
317
139
389
367
285
141
285
48
30
134
175
51
324
60
142
250
74
286
138
408
151
360
152
18
146
199
20
201
157
360
240
1
283
381
418
328
20
221
60
311
206
18
59
175
314
106
368
225
317
157
6
168
339
286
419
27
186
83
178
378
286
363
197
404
35
134
19
288
93
375
146
387
211
395
22
47
92
379
264
300
51
302
320
250
369
88
123
87
376
393
132
290
276
308
227
139
190
285
145
13
171
82
267
235
166
66
355
4

388
303
263
254
344
304
2
303
321
251
134
112
409
230
83
149
16
142
170
5
368
52
50
18
365
406
139
71
363
114
186
112
52
299
255
348
41
311
358
136
223
219
217
232
248
8
256
420
173
141
275
232
323
173
260
370
200
341
13
15
103
121
135
85
141
229
59
228
268
2
317
6
400
212
97
317
208
385
156
93
177
205
327
348
224
264
161
402
251
208
198
407
371
9
35
247
173
19
373
178
29
331
261
245
293
55
9
271
140
261
122
43
290
247
250
69
269
48
396
11
410
18
91
334
298
152
366
125
380
115
371
15
400
151
270
263
346
314
80
236
346
196
190
288
97
356
341
191
268
349
2
418
217
196
267
132
34
285
128
20
141
333
404
153
253
170
391
27
414
164
294
322
116
119
166
17
412
192
273
110
245
253
82
141
126
106
47
127
101
25
197
204
130
159
71
78
381
273
16
176
129
187
388
371
239
174
317
167
30
384
208
13
166
104
159
311
225
104
65
412
160
67
290
144
177
396
4
49
284
129
360
295
374
250
419
168
52
323
288
252
169
104
238
212
108
341
321
243
29
318
125
146
270
312
364
283
5
378
302
349
247
274
314
209
124
246


294
170
385
296
289
312
68
15
220
364
417
5
242
407
243
59
92
263
203
331
146
114
350
408
69
177
255
280
393
7
365
297
177
146
243
24
343
400
4
169
266
204
216
237
361
246
409
141
232
138
156
257
63
137
402
197
99
346
29
12
160
184
111
173
327
329
304
310
305
283
337
145
74
111
326
249
140
16
8
374
64
73
372
25
178
13
375
50
67
387
134
18
10
126
264
224
82
247
407
174
90
106
294
326
50
146
292
26
186
148
384
41
330
139
50
145
340
238
328
86
380
226
11
147
315
127
149
55
309
80
64
211
163
275
113
225
206
260
14
247
153
153
356
243
223
145
311
23
140
12
102
190
309
272
231
188
216
76
245
361
196
256
348
411
323
11
108
228
144
163
60
235
114
175
30
19
254
259
393
213
198
285
230
83
15
156
195
184
12
332
370
221
331
179
20
226
62
299
152
392
341
398
349
119
315
3
264
313
31
352
311
19
321
260
385
226
278
2
305
266
25
393
346
63
83
109
259
15
106
310
98
325
334
191
104
192
89
351
383
299
233
331
244
197
253
382
362
167
204
38
96
271
79
287
20
121
236
113
401
98
357
131
65
79
341
355
173
384

299
331
109
420
395
56
25
23
91
160
18
346
59
113
73
115
383
63
25
217
123
186
48
172
244
305
170
193
297
60
147
315
226
99
83
137
36
322
127
228
294
315
67
247
295
1
159
84
396
39
228
7
138
81
161
173
194
393
194
380
113
44
419
350
215
72
276
294
100
184
203
4
68
54
273
212
130
128
126
137
243
7
335
260
143
290
71
133
303
144
113
53
416
73
385
10
391
268
51
98
50
202
71
195
321
4
258
180
87
267
197
81
72
116
156
132
419
278
323
43
47
204
174
341
279
256
261
207
209
338
177
288
102
99
226
21
29
222
22
262
164
249
412
288
197
419
109
358
181
140
37
155
224
202
294
391
68
61
410
110
168
1
245
366
56
240
26
195
247
258
134
380
59
278
51
138
318
315
8
133
155
119
207
114
85
227
188
311
275
12
331
241
238
53
358
301
333
294
313
78
389
325
145
237
340
291
68
329
311
24
24
94
353
334
26
278
394
403
272
68
267
147
18
348
412
168
166
50
93
217
90
96
4
288
308
372
335
34
135
351
294
348
97
369
12
53
286
275
226
400
172
16
337
13
161
14
134
382
104
299
353
103
49
263
273
346
249
246
375
54
330
37

343
417
349
163
384
271
79
264
359
318
72
82
61
300
159
313
287
419
43
93
344
195
244
416
274
416
273
399
370
56
328
146
163
226
319
263
243
350
352
305
335
185
25
405
355
262
96
276
74
376
91
263
30
247
337
56
124
171
252
369
144
268
221
374
36
252
191
395
104
204
208
353
419
249
306
221
409
259
269
87
146
93
244
270
31
315
28
301
229
153
31
248
301
315
187
158
41
66
344
182
138
290
341
318
120
27
388
183
76
203
214
103
166
215
227
109
97
195
72
307
205
193
233
390
169
315
51
131
413
355
395
313
245
268
232
315
409
120
330
204
329
66
146
274
48
131
117
253
153
196
67
378
111
270
67
167
292
265
106
197
15
52
350
289
139
381
134
173
82
146
238
186
400
42
302
163
246
72
241
128
333
345
94
288
320
280
245
161
377
297
147
139
401
242
114
60
291
397
249
313
212
2
300
194
106
197
291
365
195
250
116
335
207
343
185
408
139
306
287
108
417
339
120
169
86
234
327
115
275
91
325
158
360
273
254
138
174
375
295
384
411
135
40
221
296
215
388
390
319
30
343
158
357
208
150
406
197
360
413
83
361


329
21
378
326
150
235
16
290
221
266
102
141
339
65
121
142
45
208
15
91
68
186
261
278
78
368
345
331
223
386
361
304
26
256
32
74
124
189
344
209
129
10
224
297
16
16
289
205
378
55
290
205
336
139
12
111
338
13
323
357
337
25
282
289
390
141
385
198
319
135
383
376
323
148
184
47
369
233
248
18
362
290
377
45
298
50
298
324
334
244
178
246
48
80
150
111
179
223
405
284
229
148
121
218
208
307
364
218
213
184
199
261
10
201
141
333
104
387
106
51
126
116
376
217
337
273
36
61
112
205
306
295
373
91
179
242
12
383
110
61
54
17
119
246
203
350
20
39
295
125
230
180
301
203
167
27
159
23
219
161
266
355
90
218
202
252
15
258
147
231
384
98
314
33
96
209
24
216
112
261
392
297
84
186
338
123
221
173
394
44
31
363
239
162
159
259
66
211
394
300
325
411
121
52
242
300
247
337
321
21
358
163
40
137
256
145
189
90
110
410
21
274
361
100
137
214
130
18
367
275
20
396
391
183
3
254
172
204
241
184
130
408
209
97
410
161
212
267
333
270
338
183
283
343
230
127
219
38
75
212
72
21
151
306
30
12

138
264
102
387
393
57
150
330
164
45
43
279
338
400
335
19
270
332
187
83
31
374
5
236
267
210
315
206
221
365
244
297
46
239
92
264
397
294
121
416
52
54
4
71
388
19
181
255
52
250
389
196
202
19
125
134
128
4
368
395
96
118
252
326
45
39
378
239
333
108
404
92
251
209
194
260
59
88
216
49
181
86
2
83
257
143
86
268
239
7
306
382
264
52
11
194
316
318
248
185
211
210
225
314
342
28
62
380
149
359
81
38
260
370
155
301
152
339
206
311
216
352
300
242
298
321
190
234
68
198
101
161
159
38
51
54
336
218
3
176
170
87
346
147
148
147
54
232
71
67
295
40
322
117
175
7
153
37
40
362
209
130
199
118
234
411
183
108
283
67
14
350
349
199
345
71
398
332
240
149
242
10
309
163
123
139
197
122
386
406
142
372
136
3
130
9
222
46
141
357
339
239
157
95
403
79
311
357
341
365
200
63
389
87
328
310
255
217
54
346
156
195
236
131
404
22
300
364
380
376
34
33
186
189
194
132
340
101
168
83
197
198
78
234
301
234
393
279
220
101
121
185
222
115
323
189
22
384
195
357
152
336
362
31
135
121
118
200
47
3

229
111
78
257
226
241
17
98
134
21
108
197
265
324
402
404
414
261
387
236
79
224
104
389
35
129
104
336
56
302
166
388
420
268
5
176
86
114
65
307
375
167
396
413
86
392
400
145
394
414
52
215
184
319
357
215
338
83
327
117
24
124
250
328
303
77
169
324
34
263
130
115
66
246
393
401
238
179
99
353
244
316
133
415
371
398
46
178
403
223
372
177
39
138
153
181
254
251
11
266
357
92
81
220
59
395
64
129
375
310
243
333
71
318
253
128
195
205
66
420
169
266
1
98
417
286
172
365
30
327
238
270
209
386
264
373
263
156
401
45
290
98
86
57
382
314
253
3
190
329
389
196
192
239
125
29
380
171
91
150
2
251
402
251
180
420
188
318
21
40
199
128
127
412
150
399
271
229
195
321
5
269
296
392
143
27
117
370
305
281
259
226
300
373
339
87
32
59
149
409
280
248
175
285
63
247
179
156
340
260
291
226
395
67
90
7
227
370
350
356
155
398
270
379
58
229
322
50
375
157
210
368
199
284
414
111
397
196
27
50
317
352
176
397
205
2
157
316
274
289
389
255
273
362
177
325
219
158
159
205
420
53
335
38
394
356

270
402
251
188
46
211
104
330
72
347
258
145
394
304
288
42
162
70
342
42
92
406
165
116
178
370
297
387
109
340
290
232
106
350
144
318
366
351
33
179
224
114
93
107
359
389
369
174
359
85
274
353
253
308
328
186
64
259
80
144
196
163
262
199
192
336
10
201
249
65
234
32
312
112
415
15
380
84
249
65
316
332
365
45
362
184
239
308
290
121
378
251
248
164
271
78
255
183
354
323
217
206
374
265
222
115
208
145
306
296
378
384
352
199
326
420
205
382
195
52
203
218
104
166
247
304
119
195
295
87
353
197
331
407
210
420
145
266
196
281
11
228
154
41
224
384
214
202
305
308
137
84
35
64
99
79
401
236
310
375
372
391
56
244
413
321
2
167
87
149
297
117
217
390
191
107
34
405
170
42
243
371
360
83
194
271
10
189
155
198
304
194
415
381
284
215
297
273
390
48
188
395
177
224
392
92
58
239
371
399
214
129
132
403
155
160
129
255
55
291
90
106
161
403
135
296
388
106
369
232
91
8
45
204
193
418
139
186
402
337
203
188
23
186
283
224
215
405
346
398
258
159
127
8
260
51
337
246
54
40
159
394
103

45
242
69
42
318
115
151
99
410
213
181
287
356
20
239
233
236
10
273
322
142
375
171
408
355
353
90
123
25
306
314
106
33
137
221
114
102
30
135
299
399
115
95
229
313
278
341
410
44
396
412
70
339
91
133
318
374
209
243
369
291
260
194
85
113
315
210
332
338
169
56
264
281
334
395
209
227
35
197
116
297
203
319
308
205
230
221
117
98
97
160
414
162
209
328
389
279
356
19
283
296
81
111
299
94
200
257
150
159
143
97
303
17
271
17
268
67
342
273
18
202
79
211
331
114
169
390
54
239
132
170
315
312
122
322
290
14
65
220
147
74
38
405
175
87
282
42
3
145
35
64
1
239
269
374
291
313
214
27
182
307
165
166
49
203
265
301
137
174
122
111
254
40
205
325
254
339
46
173
36
237
291
339
244
286
62
1
74
374
305
299
343
81
48
395
284
165
376
190
13
66
325
118
152
132
400
162
83
214
132
192
339
231
95
134
21
125
238
15
337
317
152
401
144
168
151
265
284
207
106
153
260
29
125
136
198
186
204
248
268
365
145
316
88
331
196
349
303
120
71
72
14
62
334
279
154
347
200
352
189
209
77
228
373
375
101
3

314
215
380
349
175
313
355
367
314
285
83
28
55
413
251
175
240
6
63
324
78
319
40
378
138
17
60
208
97
117
17
175
80
172
113
40
115
116
41
159
247
176
119
33
346
80
392
279
276
14
38
395
36
265
347
382
139
299
183
258
343
73
244
298
332
180
360
145
342
15
80
225
1
299
417
66
161
291
109
413
217
158
404
13
65
139
233
13
233
4
3
374
101
156
110
220
283
372
349
218
73
68
15
154
186
101
222
269
316
338
125
180
348
14
245
223
312
91
317
102
413
371
170
413
178
402
260
376
146
59
203
177
74
341
242
82
372
97
74
29
314
268
390
331
259
4
21
89
232
162
169
153
418
300
148
45
276
218
317
267
71
310
414
222
409
389
294
93
192
236
252
15
344
121
209
346
356
365
88
298
106
119
383
243
4
357
91
59
249
281
415
311
104
66
16
3
144
59
264
75
209
127
261
35
258
267
212
362
290
323
7
17
360
296
253
73
314
117
71
203
332
242
183
165
21
298
3
209
28
150
311
97
106
224
397
38
244
8
311
318
106
10
32
205
412
51
135
398
291
368
388
7
364
347
323
96
241
27
414
295
67
394
332
3
405
68
329
190
28
356
112
194
1

43
358
109
48
198
397
20
36
255
210
380
50
152
413
172
56
215
196
406
191
157
91
126
353
276
191
160
36
304
237
81
1
420
386
376
355
329
321
234
386
259
343
131
369
108
35
270
228
372
141
260
223
318
75
161
238
125
413
220
216
345
50
25
135
373
7
225
18
286
338
30
308
217
387
376
91
353
412
269
332
164
397
385
92
264
74
68
49
246
231
301
247
281
21
274
326
14
99
391
319
410
398
186
190
331
81
416
383
137
43
177
394
240
409
229
298
387
214
160
53
219
216
67
276
204
75
399
88
327
313
191
73
416
293
98
280
392
296
223
375
139
188
192
100
229
225
78
141
268
411
417
380
355
342
401
208
310
96
105
304
54
141
32
197
268
418
370
5
43
398
251
61
68
67
341
270
200
126
266
202
196
243
59
217
160
350
270
273
5
369
386
112
55
346
354
191
271
146
351
90
332
54
96
117
264
23
64
119
162
66
261
162
344
371
372
401
336
197
266
327
151
93
291
44
120
97
414
19
29
120
104
149
182
309
258
177
401
362
99
312
366
7
292
385
346
340
364
390
59
246
274
44
394
271
307
19
136
208
167
65
16
102
294
277
73
350
254
3

242
10
122
78
124
218
110
178
211
168
267
320
338
294
127
16
253
163
16
106
362
372
164
277
368
93
247
286
349
132
110
102
129
99
256
222
370
85
365
157
30
299
151
81
246
156
7
65
361
31
410
269
214
341
333
210
189
401
410
408
191
382
14
385
226
108
321
264
294
275
257
139
83
220
332
130
418
300
320
239
191
394
17
291
14
163
9
149
219
8
3
106
43
345
82
101
232
372
216
210
98
276
218
327
31
309
206
363
241
251
306
121
128
259
118
90
261
359
210
19
262
187
230
57
106
208
315
26
174
331
116
344
78
403
330
369
7
278
402
164
352
190
364
336
41
49
191
195
407
204
111
60
394
105
310
48
336
379
273
374
163
200
332
72
1
215
79
382
296
86
326
364
45
377
73
156
321
310
183
326
60
74
404
389
64
190
419
31
215
239
294
155
311
405
139
385
329
64
137
119
149
277
39
274
379
234
178
386
372
409
350
295
140
62
13
37
258
30
373
294
120
244
306
411
363
70
298
93
210
61
333
7
172
134
69
74
47
2
390
413
107
133
74
106
331
238
363
114
215
404
139
333
196
58
372
264
84
90
217
89
143
134
73
18
331
225
416
245


420
123
3
152
34
207
21
196
199
408
77
42
355
176
70
361
337
107
137
128
90
215
68
173
322
395
381
196
213
210
171
66
121
410
27
61
396
89
101
64
94
246
309
264
278
79
35
404
140
282
116
319
215
411
269
88
66
260
246
257
369
400
148
411
338
273
110
2
327
294
145
128
41
47
234
379
6
97
34
405
297
292
47
389
39
138
223
259
66
66
360
217
282
81
398
11
150
411
126
384
417
363
122
50
12
138
193
221
132
365
98
205
367
400
255
280
258
324
246
207
341
106
125
352
306
115
266
77
243
202
161
248
168
168
318
253
221
314
51
294
405
23
249
6
274
177
387
166
101
118
34
256
328
279
277
173
239
247
244
190
378
176
346
325
160
134
137
285
292
417
46
293
95
131
288
128
346
102
310
271
16
79
142
9
335
68
225
156
5
248
390
322
341
237
40
310
411
237
182
126
106
215
188
234
284
378
45
213
16
281
101
141
336
281
386
67
262
176
340
99
197
199
262
354
290
399
20
179
88
64
118
301
311
256
11
402
108
95
143
251
44
75
254
182
244
238
91
410
41
121
79
15
41
128
241
33
356
118
45
201
144
165
219
48
163
350
63
121


38
410
338
286
34
78
90
162
78
304
32
359
120
121
254
80
112
98
111
219
183
246
343
1
170
33
167
94
74
300
391
24
56
65
240
90
304
363
202
159
55
271
296
351
43
360
352
310
76
143
125
321
313
253
196
370
66
236
223
243
282
358
156
314
386
127
140
32
80
2
377
192
131
291
280
201
12
278
379
159
386
203
404
64
131
210
101
406
159
79
407
319
223
368
122
25
362
320
134
283
171
328
311
117
318
36
350
130
43
46
97
344
257
54
34
65
275
218
110
97
346
291
63
94
41
266
413
53
332
195
73
411
408
367
382
241
344
36
321
134
104
24
275
252
319
124
7
331
22
350
12
221
237
180
269
260
223
393
205
39
229
253
370
364
102
255
244
160
391
71
231
208
166
60
363
263
325
7
11
231
139
257
260
33
216
318
292
125
78
95
50
318
87
246
277
275
319
377
326
63
287
129
81
208
164
384
128
376
375
30
388
267
405
180
87
246
135
153
222
223
288
329
1
47
156
358
179
287
392
232
150
85
134
23
266
49
123
138
45
174
122
225
121
345
14
122
345
178
115
45
41
251
262
387
248
248
376
167
168
325
386
17
102
130
415
216
243
151
54

392
339
244
140
171
121
70
118
355
337
136
103
139
131
95
237
149
129
196
265
166
234
213
348
311
242
100
345
410
27
316
228
371
32
170
130
132
293
163
133
339
170
105
142
205
376
13
241
113
417
20
159
353
162
5
260
72
205
137
311
173
31
226
186
264
246
349
167
217
374
98
232
419
76
160
327
318
149
154
362
152
220
60
6
17
111
283
143
141
137
407
61
74
362
287
387
245
211
173
222
371
410
210
8
163
95
275
1
155
4
326
259
322
182
232
211
157
123
371
323
147
412
233
227
306
334
312
307
118
71
68
49
157
262
195
387
190
345
252
354
276
285
157
378
154
34
205
379
333
95
174
263
85
153
219
85
118
200
55
370
346
394
184
234
99
102
374
148
130
242
323
153
118
204
374
247
264
150
372
13
174
86
196
295
202
144
210
375
138
18
133
270
265
222
31
244
174
237
350
381
352
257
120
279
50
373
83
248
236
353
89
21
64
192
214
165
252
308
287
93
366
168
83
170
356
147
41
313
28
131
391
387
416
246
278
330
222
127
35
100
167
375
198
84
28
151
38
188
237
285
235
379
85
270
97
85
379
62
114
256
215
416
61
315


54
241
224
315
392
317
362
261
61
204
375
126
233
165
66
381
352
367
217
349
193
74
379
235
36
395
191
42
353
110
25
154
417
313
224
300
411
268
284
150
81
419
365
271
157
333
25
165
232
164
233
56
237
76
176
147
111
412
240
350
68
173
258
120
349
60
68
319
222
26
165
268
372
38
321
194
323
347
280
228
175
114
299
212
308
354
1
408
124
5
293
100
257
93
268
359
77
218
134
53
343
330
124
358
378
77
412
271
244
257
220
207
289
21
76
404
409
312
270
213
10
167
83
159
204
309
416
106
16
163
319
80
37
105
47
134
2
286
183
406
88
298
322
126
224
215
212
26
324
195
341
268
327
410
96
312
94
385
310
236
258
89
415
69
201
201
337
228
171
87
105
193
322
356
151
401
152
172
21
57
142
390
348
55
326
96
378
23
128
80
312
386
125
363
288
383
12
408
274
57
16
191
326
381
15
420
313
357
22
210
195
383
217
346
84
209
397
24
316
148
391
126
212
241
333
280
238
378
270
339
225
30
13
42
30
235
13
275
161
51
366
211
314
414
67
70
387
100
377
279
191
230
210
303
21
378
122
204
19
112
62
224
73
76
54
8
217
35

142
209
64
148
176
51
31
199
363
151
357
257
255
11
118
409
137
95
161
233
245
164
368
297
407
65
55
366
133
299
54
377
334
115
205
5
295
76
326
328
104
371
73
61
188
50
365
33
127
272
144
170
41
98
280
270
336
5
420
108
345
368
120
134
377
386
276
91
250
98
213
342
360
372
211
65
290
370
41
307
388
258
103
203
257
205
72
122
203
287
57
308
31
262
374
344
368
323
79
67
5
269
308
276
128
134
99
150
35
347
287
384
411
342
257
202
418
79
26
133
152
178
399
136
38
299
268
371
220
382
379
391
215
237
133
192
237
353
413
137
307
317
350
34
356
40
163
341
9
366
304
296
58
313
268
250
414
170
52
312
329
261
101
213
63
70
288
408
164
12
237
128
230
21
89
38
263
375
299
404
13
21
161
244
28
108
12
188
321
391
345
110
109
311
222
161
323
321
96
243
213
369
347
34
275
382
133
35
138
43
126
13
200
285
154
76
407
110
215
173
378
239
69
123
390
127
352
1
184
347
261
87
349
128
94
23
413
336
397
262
213
149
349
344
276
21
371
242
77
7
117
116
105
201
283
419
67
411
240
201
29
92
298
124
143
337
153
14

82
39
332
221
48
174
13
416
294
24
52
156
62
42
58
56
58
221
107
195
206
405
149
124
71
178
128
172
37
330
346
138
78
237
89
138
99
42
254
132
316
222
112
236
185
115
154
65
367
418
14
163
129
339
55
253
129
98
249
358
82
20
289
56
92
143
401
78
314
197
209
246
403
180
344
93
33
396
417
414
415
254
88
132
222
276
322
360
332
363
25
241
72
327
27
306
40
56
56
120
90
85
49
112
215
29
10
310
23
143
254
384
256
36
41
170
373
142
67
418
88
135
229
207
288
224
262
283
71
9
259
210
91
283
401
162
362
256
33
133
79
370
269
9
168
348
9
49
98
151
200
324
261
340
296
102
149
159
101
164
5
213
216
144
62
120
62
93
151
97
224
68
33
34
337
112
99
216
354
289
29
343
308
213
403
132
112
157
96
217
116
366
227
197
190
61
168
299
270
190
211
123
10
82
95
90
288
347
25
135
237
226
323
272
150
265
390
161
169
30
274
359
113
113
348
125
247
315
148
325
197
395
352
304
285
321
301
169
71
190
74
148
281
10
17
363
178
227
23
283
328
36
108
171
412
10
368
286
350
305
271
68
138
418
181
204
64
113
142
55
305
31

390
181
2
128
320
43
281
208
408
160
93
312
219
228
25
126
107
171
177
169
237
136
129
100
18
180
279
384
54
217
19
1
376
364
339
276
127
7
23
374
298
399
262
322
79
401
59
360
30
234
128
348
207
354
202
357
97
253
203
233
60
84
221
99
74
336
67
333
205
238
231
292
256
183
396
11
98
413
97
209
20
44
203
161
86
344
89
251
214
227
363
83
202
54
204
348
300
301
154
164
59
101
101
217
395
313
299
255
347
377
160
184
4
386
157
241
345
148
325
383
8
99
346
188
23
58
176
26
153
125
419
244
144
89
134
215
122
258
87
412
92
404
56
7
313
245
160
397
147
204
353
98
121
138
310
396
273
149
100
102
41
119
315
41
296
369
303
334
5
78
264
276
270
388
37
155
267
417
371
194
68
81
411
405
214
381
120
193
204
257
137
79
69
333
30
17
239
75
305
142
65
303
199
291
247
412
297
348
66
10
138
331
154
161
127
308
119
7
292
181
345
50
211
314
217
79
162
387
114
407
344
414
243
301
240
104
237
26
304
125
222
53
197
101
388
12
363
299
261
269
57
375
169
356
284
112
109
413
191
42
61
198
300
300
85
280
269
331
29

224
122
254
380
337
35
94
45
78
110
230
1
399
189
365
17
405
409
54
130
334
25
163
201
32
385
334
222
372
183
330
141
231
108
273
218
227
272
185
319
74
403
307
8
190
242
276
383
301
388
230
349
341
324
142
299
278
308
57
249
284
332
143
347
10
280
409
359
92
296
309
289
117
314
95
388
148
188
300
312
292
31
203
174
65
319
163
322
340
403
304
382
8
197
232
184
165
354
206
27
184
31
25
239
60
92
98
159
239
329
266
333
417
81
411
96
65
196
339
160
302
341
44
284
112
321
190
358
361
328
383
5
406
176
397
1
201
188
146
146
221
127
151
8
119
409
35
355
265
166
320
180
401
45
367
1
379
64
239
198
149
21
20
246
395
9
82
49
292
23
169
318
93
213
68
297
203
218
123
385
257
197
247
407
233
385
140
334
26
94
363
146
224
334
273
255
206
144
317
384
38
206
7
28
386
272
246
249
411
63
156
41
277
217
382
244
388
164
15
61
290
361
57
377
223
147
211
280
205
303
38
4
209
118
266
190
139
409
309
314
281
151
89
246
419
369
118
279
420
361
15
108
32
85
67
260
16
262
255
187
88
58
134
233
104
12
279
238
27

368
259
166
156
299
7
301
292
120
123
398
346
30
130
24
364
179
8
249
115
383
21
335
317
243
173
246
262
231
238
198
101
22
111
34
158
252
403
59
40
171
280
6
402
103
311
412
239
315
178
139
105
135
18
47
28
381
19
133
262
360
74
111
317
229
4
410
158
418
135
248
364
93
130
246
362
65
36
9
280
299
68
166
77
82
102
93
44
381
94
95
16
340
61
24
124
298
179
43
285
105
249
54
30
260
151
406
34
379
415
318
371
133
380
5
175
271
304
239
396
196
14
378
209
93
35
33
98
6
254
396
314
33
321
180
341
263
383
138
233
208
82
237
19
47
396
68
58
264
162
163
183
19
93
114
153
192
198
276
257
227
262
403
2
147
317
105
40
264
83
186
62
157
352
269
288
356
104
351
307
12
348
322
414
397
406
21
1
391
183
282
257
386
377
10
379
411
126
394
287
170
46
350
136
142
264
309
330
69
299
390
174
247
304
87
25
122
286
274
49
150
193
225
68
76
247
222
135
245
189
128
344
303
179
70
353
303
132
19
182
274
115
20
77
253
69
200
223
56
290
193
7
9
33
289
23
380
233
304
216
140
357
397
357
241
218
252
408
379
72
369
33

371
69
30
403
323
361
84
198
215
363
314
10
8
414
202
257
313
369
409
133
235
260
180
370
100
191
233
51
290
295
83
234
125
279
283
56
83
321
122
274
34
75
317
401
316
291
54
296
257
124
406
180
412
152
244
387
412
286
175
418
330
79
34
300
161
56
376
365
116
187
96
245
179
391
13
18
234
187
342
125
52
262
110
327
220
78
292
328
323
261
167
74
173
359
150
171
182
218
95
128
367
38
273
116
245
29
71
27
18
28
277
97
288
190
259
121
277
268
239
258
89
358
411
380
55
93
402
340
394
260
256
2
177
404
275
403
405
12
321
403
320
367
217
46
105
220
61
353
167
312
180
221
271
343
307
114
306
365
22
166
293
266
129
151
375
49
276
75
180
381
166
45
271
48
194
187
114
351
326
289
389
12
191
385
1
212
81
261
111
406
164
112
126
405
140
25
262
365
208
192
194
170
127
227
291
23
408
291
162
63
307
220
49
134
237
276
84
104
141
168
100
98
146
397
109
140
374
277
237
288
138
14
224
211
408
194
273
385
185
234
93
203
338
210
176
29
392
21
245
87
200
300
341
100
121
95
76
93
81
36
9
402
241
47
400
50
131

165
228
123
368
382
307
376
410
311
56
74
349
339
188
64
377
315
364
17
141
207
268
195
138
349
3
309
58
331
193
349
271
19
292
87
404
309
108
107
289
393
397
396
102
197
379
264
56
15
262
86
250
327
225
413
169
248
24
31
178
29
379
178
217
199
370
178
301
230
139
261
15
36
184
87
381
387
380
29
139
228
216
356
228
287
55
134
169
368
355
404
90
352
228
31
172
287
343
180
319
409
354
412
182
130
242
286
96
361
398
148
76
387
253
273
60
201
281
215
395
285
191
134
219
180
203
420
30
208
47
165
13
127
251
419
132
82
306
364
150
339
266
174
63
241
70
201
63
313
47
294
154
342
78
192
185
69
254
265
255
128
22
247
263
366
256
79
144
27
316
205
283
31
288
61
346
130
23
315
131
226
33
100
356
397
55
415
278
185
169
201
62
160
136
72
327
89
118
7
31
404
233
389
353
7
168
149
24
200
5
257
187
394
216
199
157
409
1
117
53
147
242
37
179
309
32
343
268
295
222
67
187
75
148
295
332
279
385
112
207
68
241
96
178
95
380
182
73
45
58
276
83
252
187
349
188
130
249
387
46
290
130
148
144
77
216
45
284

265
71
6
361
216
315
223
10
346
325
197
385
94
298
340
160
181
284
78
119
246
284
354
278
387
67
154
243
351
90
277
5
243
223
379
140
269
253
283
77
156
285
398
395
235
409
243
383
324
16
319
385
268
139
259
203
335
167
38
13
211
197
45
255
238
261
290
137
295
59
382
185
306
3
143
413
305
201
217
168
335
74
355
39
228
236
258
95
325
407
12
96
300
101
266
138
417
207
35
389
201
312
311
232
145
412
110
408
251
250
250
21
399
17
193
333
18
86
126
8
144
14
137
332
87
122
162
368
384
37
104
419
337
111
36
53
371
268
168
86
14
180
309
364
299
22
28
207
365
249
256
94
47
30
145
56
116
8
271
160
145
41
162
38
132
147
355
165
287
27
226
230
261
137
237
138
302
70
56
131
190
299
275
156
297
146
352
165
382
185
354
194
352
83
265
331
171
144
284
358
45
240
14
147
391
308
46
33
64
332
174
20
369
286
369
366
381
252
58
122
319
9
403
247
285
333
410
234
291
335
420
8
309
202
163
77
405
28
292
205
288
368
133
366
402
396
116
179
4
296
409
132
198
91
184
405
168
375
404
67
293
366
276
48
208
290
399
2

363
164
291
195
269
76
14
281
89
221
205
189
397
214
294
413
216
137
212
196
337
129
240
415
81
28
126
215
95
159
78
292
220
408
170
24
33
210
217
416
4
218
146
87
381
341
371
64
25
183
277
172
277
17
248
297
65
206
354
59
350
41
96
320
147
90
360
228
273
136
414
250
340
218
194
226
366
119
303
221
292
7
318
270
260
220
113
66
297
56
29
190
19
206
161
15
384
14
187
364
256
204
196
233
95
299
188
293
346
29
250
88
56
51
348
359
306
175
295
287
394
179
116
332
49
401
143
262
185
115
334
4
315
344
204
266
389
166
169
5
23
43
265
312
86
351
338
8
15
270
369
212
417
253
359
196
42
324
377
189
234
373
326
26
245
385
260
346
308
276
58
123
10
98
95
249
362
107
83
308
48
410
15
163
23
33
394
115
211
404
337
133
329
346
92
400
274
152
294
135
13
33
332
387
356
396
64
269
70
307
218
247
308
381
273
341
68
71
325
225
317
179
307
388
11
234
192
305
385
74
356
201
89
51
141
94
368
265
344
349
126
242
203
93
322
315
391
43
352
323
146
376
18
206
20
381
408
23
123
309
103
353
322
349
419
231
397
174


402
36
169
126
70
37
118
175
284
316
357
329
170
239
411
8
37
194
115
225
370
279
362
330
204
277
260
78
215
140
293
268
215
23
381
96
35
353
374
391
218
21
201
327
81
83
42
176
336
244
336
7
259
50
320
230
250
262
92
58
34
144
379
58
382
271
135
289
44
233
58
226
205
21
132
379
16
22
408
209
235
85
147
279
21
267
271
96
47
190
26
53
410
166
252
264
299
321
102
30
53
127
152
311
369
278
363
207
100
76
330
35
202
108
347
217
149
371
44
399
80
398
65
257
334
126
210
395
407
59
97
99
77
65
371
273
242
385
56
96
414
323
227
197
176
15
128
152
342
329
142
18
407
202
216
270
89
9
171
136
225
124
131
351
209
304
51
335
165
262
201
104
291
249
320
181
75
245
127
156
243
132
315
136
88
233
418
256
160
247
241
232
28
96
385
34
22
233
35
330
225
208
13
38
231
91
207
110
268
219
270
367
92
42
99
257
110
381
199
139
122
141
61
305
301
175
370
18
281
388
139
297
108
171
179
80
312
265
18
378
234
118
112
90
159
222
380
336
82
227
29
117
174
33
407
346
220
350
286
48
252
409
20
167
169
257
247
359
303

403
108
168
37
397
183
85
200
237
48
29
345
202
143
277
343
342
415
397
282
255
341
287
256
345
262
418
164
407
248
8
142
351
206
329
342
72
255
64
103
4
344
302
351
24
356
330
385
364
166
154
31
407
323
93
35
155
278
124
228
19
153
179
167
377
121
328
400
216
396
263
333
189
6
63
224
256
14
112
306
313
303
233
118
64
244
192
99
289
208
222
28
253
357
393
187
361
164
155
3
419
370
138
348
340
189
135
181
114
376
414
251
171
412
55
59
408
45
42
277
313
132
357
298
229
417
279
224
134
87
241
154
92
246
166
167
193
198
205
339
109
162
5
123
396
353
50
259
112
20
135
105
388
151
92
202
43
383
45
146
94
286
15
134
315
71
242
142
195
253
34
345
51
316
271
199
268
5
158
137
148
33
63
150
281
270
37
122
160
387
164
45
126
21
210
58
380
404
395
9
318
49
159
184
312
1
386
300
179
315
155
77
359
402
131
331
80
328
364
263
104
387
95
410
122
28
47
370
164
14
20
401
94
267
297
63
269
369
270
417
158
157
306
71
48
362
405
116
388
142
349
41
401
300
88
177
11
116
372
262
398
305
88
409
221
29
169
392

167
204
210
74
55
259
280
263
265
246
178
406
373
119
246
121
419
302
377
348
271
47
272
313
404
6
263
168
345
265
78
315
222
374
89
107
358
169
282
20
73
270
335
351
379
41
324
92
254
108
253
35
371
351
202
11
165
116
359
252
15
205
254
103
276
307
52
212
47
189
165
290
29
290
141
62
103
223
176
285
111
347
103
65
171
136
190
11
136
373
76
410
37
310
361
215
377
370
14
346
355
25
155
347
41
98
107
235
108
55
392
75
280
360
265
350
273
400
4
83
91
388
163
172
292
259
82
40
137
22
54
52
16
260
235
32
26
241
124
113
68
213
11
342
353
223
58
376
409
20
250
245
306
344
347
145
18
50
13
124
295
258
118
333
304
344
381
114
78
186
362
123
96
99
376
167
328
312
180
169
287
164
293
129
238
120
182
222
413
152
207
26
121
214
34
150
38
154
78
184
95
23
98
309
136
284
36
135
202
153
205
45
98
51
181
237
410
385
114
117
165
349
242
117
4
376
146
168
180
199
230
50
131
225
128
197
417
35
34
64
197
82
353
90
184
54
1
206
66
380
335
88
409
304
114
219
200
128
366
240
295
99
386
108
87
168
163
248
260


374
352
388
361
37
204
105
348
32
18
81
261
97
150
127
358
345
191
375
123
213
102
297
36
211
137
64
267
153
285
262
353
235
63
230
124
52
389
344
93
391
167
284
335
150
13
97
317
239
147
181
116
413
290
141
405
346
343
17
345
87
22
336
419
155
118
197
11
37
200
364
272
329
341
197
186
258
132
179
238
13
151
329
51
245
72
296
408
115
329
24
169
325
236
280
16
148
153
356
75
11
376
120
117
201
324
395
4
344
211
103
154
49
233
25
153
40
245
83
84
9
102
301
153
240
27
159
235
3
417
199
278
406
410
219
360
303
253
316
5
22
393
359
143
190
96
413
355
336
192
183
109
92
94
213
129
337
368
220
52
404
1
65
110
57
308
241
235
38
268
360
75
18
373
364
73
405
156
327
97
7
136
22
302
295
301
415
21
15
108
181
397
379
135
208
142
36
195
235
55
312
159
107
34
36
362
182
97
117
150
356
300
73
354
220
135
205
161
66
114
191
315
310
54
316
181
364
149
115
403
244
155
62
186
88
54
368
226
110
267
135
119
133
80
210
371
75
267
314
404
232
112
196
74
82
410
200
84
414
50
142
211
171
417
31
310
306
11
77
2

176
22
16
238
397
2
245
155
272
331
301
375
269
328
114
392
67
20
142
4
194
104
119
65
337
24
121
181
322
264
285
145
135
349
351
94
108
224
170
319
199
142
128
162
62
348
140
103
7
245
309
125
31
241
66
129
131
408
399
68
3
393
414
323
142
82
295
324
407
403
96
18
370
32
365
10
288
163
85
254
120
230
413
306
175
396
118
179
208
323
134
391
284
268
166
353
344
107
28
71
103
373
110
63
243
372
123
159
405
389
169
387
183
286
362
275
105
126
383
353
36
259
146
177
178
262
336
262
385
210
93
57
184
265
326
16
107
143
167
136
222
258
111
5
255
119
207
97
167
416
369
123
4
309
387
337
308
185
290
246
274
334
381
75
330
340
60
397
199
168
162
218
277
417
129
317
279
202
107
363
243
393
338
114
314
74
299
275
377
331
249
35
330
243
357
272
242
359
22
262
89
142
341
135
318
350
88
185
175
84
180
148
298
31
372
364
188
370
91
406
403
198
254
113
230
287
268
299
81
59
207
416
358
85
257
78
17
213
255
417
84
179
386
143
393
279
265
109
62
269
21
282
420
387
88
269
382
336
334
371
363
266
220
95
7

306
236
192
40
332
179
230
287
91
19
62
387
164
9
21
51
341
35
282
340
67
125
324
113
330
239
398
159
248
369
96
13
231
211
166
282
358
33
205
375
67
258
214
97
80
319
234
215
261
182
114
46
322
420
364
120
179
222
402
351
138
244
22
203
122
259
193
392
202
242
65
31
43
229
394
351
145
54
221
371
178
275
408
42
340
393
162
227
326
247
66
14
311
87
194
309
71
27
368
137
90
136
208
268
305
235
56
136
111
368
61
377
199
82
191
138
258
395
183
419
299
322
159
404
2
347
158
420
190
165
137
111
193
233
28
54
191
93
108
94
174
401
326
202
95
329
187
70
253
146
266
400
384
328
362
10
215
345
290
157
364
169
294
93
393
387
108
336
356
361
113
165
145
113
243
21
418
408
84
108
3
195
210
142
93
16
355
70
280
339
80
262
334
387
415
173
25
371
274
87
232
228
55
319
420
249
220
217
64
133
364
40
208
391
62
133
375
161
220
92
113
414
212
60
207
135
316
16
340
400
21
106
374
350
32
277
11
133
232
63
351
255
386
394
369
385
304
412
78
410
37
406
147
331
93
180
309
158
56
62
297
374
384
384
273
141
144


253
17
12
383
195
260
281
73
370
127
175
242
284
51
182
400
60
420
385
114
385
417
178
420
126
38
270
28
225
153
253
290
55
66
91
280
24
44
17
141
299
304
212
382
200
169
395
154
314
150
208
167
231
299
260
409
312
171
352
345
313
43
245
189
313
148
149
385
85
254
121
308
215
6
314
40
145
304
408
139
358
354
114
192
329
343
216
248
291
409
214
8
223
400
328
93
12
275
314
205
271
220
100
234
391
315
352
96
286
321
33
375
323
329
57
317
115
335
110
167
271
252
5
285
191
10
302
369
299
86
407
204
166
88
315
393
1
162
26
249
103
125
331
321
52
130
332
240
296
96
385
309
9
75
176
339
266
333
63
92
325
259
78
223
257
25
146
71
144
102
246
198
376
256
60
407
284
334
362
44
69
354
45
83
420
122
305
282
311
129
317
110
256
160
207
86
80
68
339
214
350
331
70
374
37
100
192
229
325
97
74
75
309
317
406
420
90
371
220
240
274
153
171
198
194
50
103
206
209
368
347
199
244
191
374
283
88
102
261
211
336
103
73
185
185
13
183
240
356
219
252
252
377
130
349
226
44
190
280
193
56
253
8
134
3
43
16
3

KeyboardInterrupt: 

In [10]:
random_gen = generate_random_nums()
for _ in range(10):
    print(next(random_gen))
    

307
390
91
78
374
165
100
273
103
27


In [16]:
list(range(1,10,2))

[1, 3, 5, 7, 9]

In [19]:
# Build a range interator on your own 

def my_range(start, stop, step=1):
    while start < stop:
        yield start
        start+=step
    
for x in my_range(20,50,5):
    print(x)
        

20
25
30
35
40
45


In [20]:
# Create a genator that generates infintely the factorial
def factorial_generator():
    n = 0
    factorial = 1
    while True:
        yield factorial
        n += 1
        factorial *= n
        
fg = factorial_generator()
for _ in range(10):
    print(next(fg))
        


1
1
2
6
24
120
720
5040
40320
362880


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC"]


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>
<code>fib(1) = 0
fib(2) = 1
fib(3) = 1
fib(4) = 2
fib(5) = 3
fib(6) = 5
fib(7) = 8
fib(8) = 13
fib(9) = 21
fib(10) = 34
fib(11) = 55
fib(12) = 89
fib(13) = 144
</code>

### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def my_gen(number):
    pass

for num in my_gen(10):
    print(num)
    
# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 2
# 1
# 0